### PR TITLE
Distribute GAS to voters

### LINKED ISSUE #1348: Rework GAS distribution scheme, follow economic model updates
We need to carefully review and implement already merged neo-project/neo#1845 and neo-project/neo#1847. Then move on to neo-project/neo#1848 that also won't wait for long to merged. This is quite an important change, so we need to also check if there are any problems with this scheme.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1348 .

### CODE REVIEW

**Comment by roman-khimov** [comment]
Take a look at neo-project/neo#2049 also, this will require updating a number of tests that you touch here anyway.

File: pkg/core/blockchain.go [hunk 1]
```diff
@@ -1144,7 +1144,7 @@ func (bc *Blockchain) UnsubscribeFromExecutions(ch chan<- *state.AppExecResult)
 // amount of NEO between specified blocks.
 func (bc *Blockchain) CalculateClaimable(value *big.Int, startHeight, endHeight uint32) *big.Int {
 	ic := bc.newInteropContext(trigger.Application, bc.dao, nil, nil)
-	res, _ := bc.contracts.NEO.CalculateBonus(ic, value, startHeight, endHeight)
+	res, _ := bc.contracts.NEO.CalculateNEOHolderReward(ic, value, startHeight, endHeight)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/blockchain.go:1147]: Should RPC `getclaimable` return funds including GAS from voting?

File: pkg/core/native/native_neo.go [hunk 2]
```diff
@@ -451,13 +508,58 @@ func (n *NEO) SetGASPerBlock(ic *interop.Context, index uint32, gas *big.Int) (b
...
+}
+
+// CalculateNEOHolderReward return GAS reward for holding `value` of NEO from start to end block.
+func (n *NEO) CalculateNEOHolderReward(ic *interop.Context, value *big.Int, start, end uint32) (*big.Int, error) {
 	if value.Sign() == 0 || start >= end {
 		return big.NewInt(0), nil
 	} else if value.Sign() < 0 {
 		return nil, errors.New("negative value")
 	}
+
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go]: Fixed.

File: pkg/core/native/native_neo.go [hunk 3]
```diff
@@ -231,8 +227,17 @@ func (n *NEO) InitializeCache(bc blockchainer.Blockchainer, d dao.DAO) error {
...
-func (n *NEO) updateCache(committee keys.PublicKeys, bc blockchainer.Blockchainer) error {
+func (n *NEO) updateCache(cvs keysWithVotes, bc blockchainer.Blockchainer) error {
+	var err error
+	var committee = make(keys.PublicKeys, len(cvs))
+	for i := range committee {
+		committee[i], err = cvs[i].PublicKey()
+		if err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go:242]: Fixed.

File: pkg/core/blockchain.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/blockchain.go]: Fixed.

File: pkg/core/native/native_neo.go [hunk 4]
```diff
@@ -286,13 +290,66 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
+					tmp = big.NewInt(2)
+				}
+				tmp.Mul(tmp, voterReward)
+				tmp.Div(tmp, cs[i].Votes)
+
+				var key = makeVoterKey([]byte(cs[i].Key))
+				var reward = n.getGASPerVote(ic.DAO, key, ic.Block.Index+1)
+				tmp.Add(tmp, &reward[0])
+
+				var buf = make([]byte, 4)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:321]: Better preallocate it once before all iterations.

Comment (roman-khimov) [reply pkg/core/native/native_neo.go:321]: And this is probably true for most of other local variables here.

File: pkg/core/native/native_neo.go [hunk 5]
```diff
@@ -57,6 +57,10 @@ const (
 	prefixCandidate = 33
 	// prefixVotersCount is a prefix for storing total amount of NEO of voters.
 	prefixVotersCount = 1
+	// prefixVoterRewardPerCommittee is a prefix for storing committee GAS reward.
+	prefixVoterRewardPerCommittee = 23
+	// storeFactor is a factor by which voter reward per committee is multiplied.
+	storeFactor = 100_000_000
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:64]: `voterRewardFactor`?

File: pkg/core/native/native_neo.go [hunk 6]
```diff
@@ -444,8 +501,52 @@ func (n *NEO) SetGASPerBlock(ic *interop.Context, index uint32, gas *big.Int) (b
 	return true, n.putGASRecord(ic.DAO, index, gas)
 }
 
-// CalculateBonus calculates amount of gas generated for holding `value` NEO from start to end block.
-func (n *NEO) CalculateBonus(ic *interop.Context, value *big.Int, start, end uint32) (*big.Int, error) {
+func (n *NEO) checkCandidate(d dao.DAO, pub *keys.PublicKey, c *candidate) error {
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:499]: `dropCandidateIfZero`? I mean it's not a `check` really.

File: pkg/core/native/native_neo.go [hunk 7]
```diff
@@ -444,8 +501,52 @@ func (n *NEO) SetGASPerBlock(ic *interop.Context, index uint32, gas *big.Int) (b
...
+	for i := range toRemove {
+		if err := d.DeleteStorageItem(n.ContractID, []byte(toRemove[i])); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func makeVoterKey(pub []byte) []byte {
+	key := make([]byte, 38)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:520]:
```suggestion
	key := make([]byte, 34, 38)
```
I don't think current code is correct.

File: pkg/core/native/native_neo.go [hunk 8]
```diff
@@ -600,9 +704,7 @@ func (n *NEO) ModifyAccountVotes(acc *state.NEOBalanceState, d dao.DAO, value *b
 		cd := new(candidate).FromBytes(si.Value)
 		cd.Votes.Add(&cd.Votes, value)
 		if !isNewVote {
-			if !cd.Registered && cd.Votes.Sign() == 0 {
-				return d.DeleteStorageItem(n.ContractID, key)
-			}
+			return n.checkCandidate(d, acc.VoteTo, cd)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:715]: This breaks the logic, you're not supposed to `return` if the validator is still active.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/native/native_neo.go [hunk 9]
```diff
@@ -42,18 +42,12 @@ type NEO struct {
 	// (every 28 blocks for mainnet). It's value
 	// is always equal to value stored by `prefixCommittee`.
 	committee atomic.Value
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go:45]: Yeah, forgot to remove. Fixed.

File: pkg/core/dao/dao.go [hunk 10]
```diff
@@ -422,17 +419,25 @@ func (dao *Simple) GetStorageItemsWithPrefix(id int32, prefix []byte) (map[strin
...
+// Seek executes f for all items with a given prefix.
+func (dao *Simple) Seek(id int32, prefix []byte, f func(k, v []byte)) {
+	lookupKey := makeStorageItemKey(id, nil)
+	if prefix != nil {
+		lookupKey = append(lookupKey, prefix...)
+	}
+	dao.Store.Seek(lookupKey, func(k, v []byte) {
 		// Cut prefix and hash.
 		// Must copy here, #1468.
 		key := make([]byte, len(k[len(lookupKey):]))
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/dao/dao.go:437]: Fixed.
I am now a bit confused about this copying as map keys use `string(key)` which shouldn't be affected by changing key later.

File: pkg/core/native/native_neo.go (hunk 5 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:64]: Fixed

File: pkg/core/native/native_neo.go [hunk 11]
```diff
@@ -286,13 +290,66 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
 	pubs := n.GetCommitteeMembers()
-	index := int(ic.Block.Index) % len(ic.Chain.GetConfig().StandbyCommittee)
+	committeeSize := len(ic.Chain.GetConfig().StandbyCommittee)
+	index := int(ic.Block.Index) % committeeSize
 	gas.Mul(gas, big.NewInt(committeeRewardRatio))
 	n.GAS.mint(ic, pubs[index].GetScriptHash(), gas.Div(gas, big.NewInt(100)))
+
+	if shouldUpdateCommittee(ic.Block.Index, ic.Chain) {
+		var voterReward = big.NewInt(voterRewardRatio)
+		voterReward.Mul(voterReward, gas)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go:301]: Fixed.

File: pkg/core/native/native_neo.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:321]: Fixed.

File: pkg/core/native/native_neo.go (hunk 7 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:520]: Fixed.

File: pkg/core/native/native_neo.go (hunk 6 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:499]: Fixed.

File: pkg/core/native/native_neo.go (hunk 8 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:715]: Fixed.

File: pkg/core/native/native_neo.go [hunk 12]
```diff
@@ -283,13 +288,68 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
+				}
+				tmp.Mul(tmp, voterReward)
+				tmp.Div(tmp, cs[i].Votes)
+
+				key = makeVoterKey([]byte(cs[i].Key), key[:34])
+				var reward = n.getGASPerVote(ic.DAO, key, ic.Block.Index+1)
+				tmp.Add(tmp, &reward[0])
+
+				binary.BigEndian.PutUint32(buf, ic.Block.Index+1)
+				key = append(key, buf...)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:323]: `binary.BigEndian.PutUint32(key[34:], ic.Block.Index+1)` and drop the buf/append?

File: pkg/core/native/native_neo.go [hunk 13]
```diff
@@ -283,13 +288,68 @@ func (n *NEO) OnPersist(ic *interop.Context) error {
...
+		for i := range cs {
+			if cs[i].Votes.Sign() > 0 {
+				tmp := big.NewInt(1)
+				if i < validatorsCount {
+					tmp = big.NewInt(2)
+				}
+				tmp.Mul(tmp, voterReward)
+				tmp.Div(tmp, cs[i].Votes)
+
+				key = makeVoterKey([]byte(cs[i].Key), key[:34])
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:318]: `[:34]` would need to be removed here, but it should work.

File: pkg/core/native/native_neo.go (hunk 12 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:323]: fixed.

File: pkg/core/native/native_neo.go (hunk 13 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:318]: `[:34]` is necessary either here or on the next line, because the seek should be done by 34-byte prefix.

Comment (fyrchik) [reply pkg/core/native/native_neo.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Comment by roman-khimov** [comment]
```
    wallet_test.go:209: 
        	Error Trace:	wallet_test.go:209
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestClaimGas
```
Why does it happen?

**Comment by fyrchik** [comment]
@roman-khimov I am investigating. It happens with `-race` flag. Most likely, the problem is that we can't determine height when claim tx will be persisted, but this height is needed to calculate expected amount of GAS claimed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**